IllegalAccessError when using OpenTelemetry extension with custom Resource

### Describe the bug

I created a quarkus project with the OpenTelemetry extension and  added a custom `Resource CDI Producer`.

```java
import static io.opentelemetry.semconv.resource.attributes.ResourceAttributes.SERVICE_INSTANCE_ID;

import java.util.UUID;
import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.inject.Produces;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.sdk.resources.Resource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@ApplicationScoped
public class ResourceConfiguration {
    private static final Logger log = LoggerFactory.getLogger(ResourceConfiguration.class);
    
    @Produces
    @ApplicationScoped
    public Resource newRelicResource() {
        log.info("Adding custom resources {}", ResourceConfiguration.class.getName());
        return Resource.create(
            Attributes.builder().put(SERVICE_INSTANCE_ID, UUID.randomUUID().toString()).build());
    }
}
```

### Expected behavior

The application should startup when `./mvnw clean quarkus:dev`

### Actual behavior

An exception is thrown

```posh
2022-01-11 16:39:21,512 DEBUG [org.hib.val.int.uti.Version] (Quarkus Main Thread) HV000001: Hibernate Validator 6.2.1.Final
2022-01-11 16:39:22,852 ERROR [io.qua.run.Application] (Quarkus Main Thread) Failed to start application (with profile dev): java.lang.IllegalAccessError: class org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_ClientProxy tried to access method 'void io.opentelemetry.sdk.resources.Resource.<init>()' (org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_ClientProxy is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @cd7f1ae; io.opentelemetry.sdk.resources.Resource is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @76b10754)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_ClientProxy.<init>(Unknown Source)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_Bean.proxy(Unknown Source)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_Bean.get(Unknown Source)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_Bean.get(Unknown Source)
        at io.quarkus.arc.impl.InstanceImpl.getBeanInstance(InstanceImpl.java:223)
        at io.quarkus.arc.impl.InstanceImpl$InstanceIterator.next(InstanceImpl.java:259)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:657)
        at io.quarkus.opentelemetry.runtime.tracing.TracerRecorder.setupResources(TracerRecorder.java:117)
        at io.quarkus.deployment.steps.TracerProcessor$setupTracer558615094.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.TracerProcessor$setupTracer558615094.deploy(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:104)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:67)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:120)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:103)
        at java.base/java.lang.Thread.run(Thread.java:833)


2022-01-11 16:39:22,852 ERROR [io.qua.run.Application] (Quarkus Main Thread) Failed to start application (with profile dev): java.lang.IllegalAccessError: class org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_ClientProxy tried to access method 'void io.opentelemetry.sdk.resources.Resource.<init>()' (org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_ClientProxy is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @cd7f1ae; io.opentelemetry.sdk.resources.Resource is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @76b10754)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_ClientProxy.<init>(Unknown Source)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_Bean.proxy(Unknown Source)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_Bean.get(Unknown Source)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_Bean.get(Unknown Source)
        at io.quarkus.arc.impl.InstanceImpl.getBeanInstance(InstanceImpl.java:223)
        at io.quarkus.arc.impl.InstanceImpl$InstanceIterator.next(InstanceImpl.java:259)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:657)
        at io.quarkus.opentelemetry.runtime.tracing.TracerRecorder.setupResources(TracerRecorder.java:117)
        at io.quarkus.deployment.steps.TracerProcessor$setupTracer558615094.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.TracerProcessor$setupTracer558615094.deploy(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:104)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:67)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:120)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:103)
        at java.base/java.lang.Thread.run(Thread.java:833)


2022-01-11 16:39:22,853 INFO  [io.qua.dep.dev.IsolatedDevModeMain] (main) Attempting to start live reload endpoint to recover from previous Quarkus startup failure
2022-01-11 16:39:22,853 INFO  [io.qua.dep.dev.IsolatedDevModeMain] (main) Attempting to start live reload endpoint to recover from previous Quarkus startup failure
2022-01-11 16:39:22,913 INFO  [org.jbo.threads] (main) JBoss Threads version 3.4.2.Final
2022-01-11 16:39:23,002 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.dev.appstate.ApplicationStateNotification.waitForApplicationStart(ApplicationStateNotification.java:51)
        at io.quarkus.runner.bootstrap.StartupActionImpl.runMainClass(StartupActionImpl.java:122)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:145)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:456)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:67)
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:149)
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:105)
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:145)
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:63)
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:104)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:67)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:120)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:103)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalAccessError: class org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_ClientProxy tried to access method 'void io.opentelemetry.sdk.resources.Resource.<init>()' (org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_ClientProxy is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @cd7f1ae; io.opentelemetry.sdk.resources.Resource is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @76b10754)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_ClientProxy.<init>(Unknown Source)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_Bean.proxy(Unknown Source)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_Bean.get(Unknown Source)
        at org.acme.otel.ResourceConfiguration_ProducerMethod_newRelicResource_625b8ffec6a326c4440bff06cc0a52e42734afec_Bean.get(Unknown Source)
        at io.quarkus.arc.impl.InstanceImpl.getBeanInstance(InstanceImpl.java:223)
        at io.quarkus.arc.impl.InstanceImpl$InstanceIterator.next(InstanceImpl.java:259)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:657)
        at io.quarkus.opentelemetry.runtime.tracing.TracerRecorder.setupResources(TracerRecorder.java:117)
        at io.quarkus.deployment.steps.TracerProcessor$setupTracer558615094.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.TracerProcessor$setupTracer558615094.deploy(Unknown Source)
        ... 13 more
```

### How to Reproduce?

[reproducer.zip](https://github.com/quarkusio/quarkus/files/7849254/reproducer.zip)


### Output of `uname -a` or `ver`

Linux usename 5.10.60.1-microsoft-standard-WSL2 

### Output of `java -version`

openjdk version "17.0.1" 2021-10-19 OpenJDK Runtime Environment (build 17.0.1+12-Ubuntu-120.04) OpenJDK 64-Bit Server VM (build 17.0.1+12-Ubuntu-120.04, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.6.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d) Maven home:  Java version: 17.0.1, vendor: Private Build, runtime: /usr/lib/jvm/java-17-openjdk-amd64 Default locale: en, platform encoding: UTF-8 OS name: "linux", version: "5.10.60.1-microsoft-standard-wsl2", arch: "amd64", family: "unix"

### Additional information

_No response_

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/ProducerReturnTypePackagePrivateNoArgsConstructorTest.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/ProducerReturnTypePackagePrivateNoArgsConstructorTest.java
@@ -1,0 +1,50 @@
+package io.quarkus.arc.test.unproxyable;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.io.IOException;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Instance;
+import javax.enterprise.inject.Produces;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.test.unproxyable.some.Resource;
+import io.quarkus.test.QuarkusUnitTest;
+
+// This test aims to test the https://github.com/quarkusio/quarkus/issues/22815 in Quarkus integration
+// There is a duplicate test for ArC standalone: io.quarkus.arc.test.clientproxy.constructor.ProducerReturnTypePackagePrivateNoArgsConstructorTest
+public class ProducerReturnTypePackagePrivateNoArgsConstructorTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot(root -> root
+                    .addClasses(ProducerReturnTypePackagePrivateNoArgsConstructorTest.class, ResourceProducer.class,
+                            Resource.class));
+
+    @Inject
+    Instance<Resource> instance;
+
+    @Test
+    public void testProducer() throws IOException {
+        assertTrue(instance.isResolvable());
+        assertEquals(5, instance.get().ping());
+    }
+
+    @Singleton
+    static class ResourceProducer {
+
+        @ApplicationScoped
+        @Produces
+        Resource resource() {
+            return Resource.from(5);
+        }
+
+    }
+
+}

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/some/Resource.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/some/Resource.java
@@ -1,0 +1,20 @@
+package io.quarkus.arc.test.unproxyable.some;
+
+public abstract class Resource {
+
+    Resource() {
+    }
+
+    public static Resource from(int ping) {
+        return new Resource() {
+
+            @Override
+            public int ping() {
+                return ping;
+            }
+        };
+    }
+
+    public abstract int ping();
+
+}

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
@@ -895,7 +895,7 @@ public class BeanDeployment {
 
             // non-inherited stuff:
             for (MethodInfo method : beanClass.methods()) {
-                if (Methods.isSynthetic(method)) {
+                if (method.isSynthetic()) {
                     continue;
                 }
                 if (annotationStore.getAnnotations(method).isEmpty()) {
@@ -922,7 +922,7 @@ public class BeanDeployment {
             while (aClass != null) {
                 for (MethodInfo method : aClass.methods()) {
                     Methods.MethodKey methodDescriptor = new Methods.MethodKey(method);
-                    if (Methods.isSynthetic(method) || Methods.isOverriden(methodDescriptor, methods)) {
+                    if (method.isSynthetic() || Methods.isOverriden(methodDescriptor, methods)) {
                         continue;
                     }
                     methods.add(methodDescriptor);

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java
@@ -1793,7 +1793,7 @@ public class BeanGenerator extends AbstractGenerator {
 
     private String getProxyTypeName(BeanInfo bean, String baseName) {
         StringBuilder proxyTypeName = new StringBuilder();
-        proxyTypeName.append(bean.getTargetPackageName());
+        proxyTypeName.append(bean.getClientProxyPackageName());
         if (proxyTypeName.length() > 0) {
             proxyTypeName.append(".");
         }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
@@ -492,6 +492,22 @@ public class BeanInfo implements InjectionTargetInfo {
         return packageName;
     }
 
+    public String getClientProxyPackageName() {
+        if (isProducerField() || isProducerMethod()) {
+            AnnotationTarget target = getTarget().get();
+            DotName typeName = target.kind() == Kind.FIELD ? target.asField().type().name()
+                    : target.asMethod().returnType().name();
+            String packageName = DotNames.packageName(typeName);
+            if (packageName.startsWith("java.")) {
+                // It is not possible to place a class in a JDK package
+                packageName = AbstractGenerator.DEFAULT_PACKAGE;
+            }
+            return packageName;
+        } else {
+            return getTargetPackageName();
+        }
+    }
+
     void validate(List<Throwable> errors, List<BeanDeploymentValidator> validators,
             Consumer<BytecodeTransformer> bytecodeTransformerConsumer, Set<DotName> classesReceivingNoArgsCtor) {
         Beans.validateBean(this, errors, validators, bytecodeTransformerConsumer, classesReceivingNoArgsCtor);

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
@@ -84,7 +84,7 @@ public class ClientProxyGenerator extends AbstractGenerator {
         ProviderType providerType = new ProviderType(bean.getProviderType());
         ClassInfo providerClass = getClassByName(bean.getDeployment().getBeanArchiveIndex(), providerType.name());
         String baseName = getBaseName(bean, beanClassName);
-        String targetPackage = bean.getTargetPackageName();
+        String targetPackage = bean.getClientProxyPackageName();
         String generatedName = generatedNameFromTarget(targetPackage, baseName, CLIENT_PROXY_SUFFIX);
         if (existingClasses.contains(generatedName)) {
             return Collections.emptyList();

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
@@ -42,10 +42,9 @@ final class Methods {
     public static final String INIT = "<init>";
     // static initializer
     public static final String CLINIT = "<clinit>";
-    // copied from java.lang.reflect.Modifier.SYNTHETIC
-    static final int SYNTHETIC = 0x00001000;
     // copied from java.lang.reflect.Modifier.BRIDGE
     static final int BRIDGE = 0x00000040;
+
     public static final String TO_STRING = "toString";
 
     private static final List<String> IGNORED_METHODS = initIgnoredMethods();
@@ -58,10 +57,6 @@ final class Methods {
     }
 
     private Methods() {
-    }
-
-    static boolean isSynthetic(MethodInfo method) {
-        return (method.flags() & SYNTHETIC) != 0;
     }
 
     static boolean isBridge(MethodInfo method) {

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/constructor/ProducerReturnTypePackagePrivateNoArgsConstructorTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/constructor/ProducerReturnTypePackagePrivateNoArgsConstructorTest.java
@@ -1,0 +1,44 @@
+package io.quarkus.arc.test.clientproxy.constructor;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.ClientProxy;
+import io.quarkus.arc.test.ArcTestContainer;
+import io.quarkus.arc.test.clientproxy.constructor.some.Resource;
+import java.io.IOException;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+import javax.inject.Singleton;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+// This test aims to test the https://github.com/quarkusio/quarkus/issues/22815 in ArC standalone
+// There is a duplicate test for Quarkus integration: io.quarkus.arc.test.unproxyable.ProducerReturnTypePackagePrivateNoArgsConstructorTest
+public class ProducerReturnTypePackagePrivateNoArgsConstructorTest {
+
+    @RegisterExtension
+    public ArcTestContainer container = new ArcTestContainer(ResourceProducer.class);
+
+    @Test
+    public void testProducer() throws IOException {
+        Resource res = Arc.container().instance(Resource.class).get();
+        assertNotNull(res);
+        assertTrue(res instanceof ClientProxy);
+        assertEquals(5, res.ping());
+    }
+
+    @Singleton
+    static class ResourceProducer {
+
+        @ApplicationScoped
+        @Produces
+        Resource resource() {
+            return Resource.from(5);
+        }
+
+    }
+
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/constructor/some/Resource.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/constructor/some/Resource.java
@@ -1,0 +1,20 @@
+package io.quarkus.arc.test.clientproxy.constructor.some;
+
+public abstract class Resource {
+
+    Resource() {
+    }
+
+    public static Resource from(int ping) {
+        return new Resource() {
+
+            @Override
+            public int ping() {
+                return ping;
+            }
+        };
+    }
+
+    public abstract int ping();
+
+}

